cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -124,44 +144,212 @@ def resolve_packages(source_dir: RootedPath) -> list[Package]:
+            if not package.cache_path:
+                raise _CouldNotResolve(
+                    "expected a zip archive in the cache but 'yarn info' says there is none",
+                )
+            cache_path = self._cache_path_as_rooted(package.cache_path)
+            if not cache_path.path.exists():
+                raise _CouldNotResolve(
+                    f"cache archive does not exist: {cache_path.subpath_from_root}"
+                )
+            log_for_locator("reading package name from %s", cache_path.subpath_from_root)
+            # file and https dependencies have reliable versions but unreliable names
+            name = self._read_name_from_cache(cache_path)
+            version = package.version
+        elif isinstance(locator, (PortalLocator, LinkLocator)):
+            parent_locator = locator.locator
+            packjson = self._project_subpath(
+                parent_locator.relpath, locator.relpath, "package.json"
+            )
+            if isinstance(locator, LinkLocator) and not packjson.path.exists():
+                # if a link dependency doesn't have a package.json, we have to rely on the locator
+                name = self._scoped_name(locator)
+                version = None
+            else:
+                # otherwise, take both the name and the version from package.json
+                # (name is unreliable, version is '0.0.0-use.local')
+                log_for_locator(
+                    "reading package name and version from %s", packjson.subpath_from_root
+                )
+                name, version = self._read_name_version_from_packjson(packjson)
+        elif isinstance(locator, PatchLocator):
+            if (
+                package.cache_path
+                # yarn info seems to always report the cache path for patch dependencies,
+                # but the path doesn't always exist
+                and (cache_path := self._cache_path_as_rooted(package.cache_path)).path.exists()
+            ):
+                log_for_locator("reading package name from %s", cache_path.subpath_from_root)
+                name = self._read_name_from_cache(cache_path)
+            elif orig_package := self._package_mapping.get(locator.package):
+                log_for_locator("resolving the name of the original package")
+                name = self._resolve_package(orig_package).name
+            else:
+                raise _CouldNotResolve(
+                    "the 'yarn info' output does not include either an existing zip archive "
+                    "or the original unpatched package",
+                )
+            version = package.version
+        else:
+            # This line can never be reached assuming type-checker checks are passing
+            # https://typing.readthedocs.io/en/latest/source/unreachable.html#assert-never-and-exhaustiveness-checking
+            assert_never(locator)
+
+        return _ResolvedPackage(locator, name, version, package.raw_locator)
```

Comment:
Seeing this changed to a single `return` makes the Functional Programmer in me very happy :D

Reply:
I just realized that CodeQL's earlier problem with that local variable`version` being "used before initialize" is because that "single exit" pattern *requires* explicitly declaring (e.g. `version = None` or maybe the correct type, probably around [here](https://github.com/chmeliik/cachi2/blob/bd4db5f7c5942f23fcb9610dd54e62cf5e9697ea/cachi2/core/package_managers/yarn/resolver.py#L208)) anything you return which *isn't* already assigned to in the "top level" scope (for whatever "top-level" is in context). I hope that made sense...